src/backend/distributed/utils/resource_lock.c

```diff
@@ -1196,3 +1216,164 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
 		SendCommandToWorker(nodeName, nodePort, lockCommand);
 	}
 }
+
+
+/*
+ * AcquireDistributedLockOnRelations filters relations before passing them to
+ * AcquireDistributedLockOnRelations_Internal to acquire the locks.
+ *
+ * Only tables, views, and foreign tables can be locked with this function. Other relations
+ * will cause an error.
+ *
+ * Skips non distributed relations.
+ * configs parameter is used to configure what relation will be locked and if the lock
+ * should throw an error if it cannot be acquired immediately
+ */
+void
+AcquireDistributedLockOnRelations(List *relationList, LOCKMODE lockMode, uint32 configs)
+{
+	if (!ClusterHasKnownMetadataWorkers() || !EnableMetadataSync || !EnableDDLPropagation)
+	{
+		return;
+	}
+
+	List *distributedRelationList = NIL;
+	List *lockDistRelDescendantsList = NIL;
+
+	/*
+	 * this is used to enforce the lock order:
+	 * [...A], [...B]
+	 * Where:
+	 * A: Explicitly locked relations
+	 * B: Relations with a foreign key relationships to relationships in A,
+	 *    which are not locked by the lock command by default
+	 *
+	 * e.g.
+	 *  Let view v with tables t1, t2
+	 *  for LOCK v IN * MODE;
+	 *  A: v
+	 *  B: nil
+	 *  the lock command will lock t1 and t2 so no need to populate B
+	 *
+	 *  Let tables t1 and t2 with t2 referencing t1 with a foreign key
+	 *  for TRUNCATE t1 (we need to acquire locks on t1 and t2)
+	 *  A: t1
+	 *  B: t2
+	 *  t2 will not be locked by lock command if we call LOCK t1
+	 *  so we need to add it to B
+	 */
+	List *referencingRelationIds = NIL;
+
+	bool nowait = (configs & DIST_LOCK_NOWAIT) > 0;
+
+	RangeVar *rangeVar = NULL;
+	foreach_ptr(rangeVar, relationList)
+	{
+		bool lockDescendants = rangeVar->inh;
+		Oid relationId = RangeVarGetRelidExtended(rangeVar, AccessShareLock, nowait ?
+												  RVR_NOWAIT : 0, NULL, NULL);
+
+		/*
+		 * Note that allowing the user to lock shards could lead to
+		 * distributed deadlocks due to shards not being locked when
+		 * a distributed table is locked.
+		 * However, because citus.enable_manual_changes_to_shards
+		 * is a guc which is not visible by default, whoever is using this
+		 * guc will hopefully know what they're doing and avoid such scenarios.
+		 */
+		ErrorIfIllegallyChangingKnownShard(relationId);
+
+		/*
+		 * we want to prevent under privileged users to trigger establishing connections,
+		 * that's why we have this additional check. Otherwise, we'd get the errors as
+		 * soon as we execute the command on any node
+		 */
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		if (get_rel_relkind(relationId) == RELKIND_VIEW)
```

Comment:
minor simplification suggestion:

I think the rest of the loop can be moved to a separate function like:
```C

List *relationsToBeLocked = FindRelationsToBeLocked(relationId);
distributedRelationList = list_concat(distributedRelationList, relationsToBeLocked)
```

In general, we try to avoid more than `3` levels of indentation (if/while/switch etc) for keeping the code relatively easier to follow